hal/reset_reason_api.h

```diff
@@ -0,0 +1,65 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_RESET_REASON_API_H
+#define MBED_RESET_REASON_API_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum
+{
+  RESET_REASON_POWER_ON  = (1 << 0),
+  RESET_REASON_PIN_RESET = (1 << 1),
+  RESET_REASON_BROWN_OUT = (1 << 2),
+  RESET_REASON_SOFTWARE  = (1 << 3),
+  RESET_REASON_WATCHDOG  = (1 << 4),
+  RESET_REASON_LOCKUP    = (1 << 5),
+  RESET_REASON_UNKNOWN   = (1 << 6)
+} reset_reason_t;
+
+// Fetch the reset reason for the last system reset
+//
+// Note: Some platforms contain reset reason registers that persist through
+// system resets. If the registers haven't been cleared before calling this
+// function multiple reasons may be set within the registers. This function will
+// return the first reason it identifies, this may not be the cause of the
+// previous reset.
+//
+// @return enum containing the last reset reason for the board. If the enum does
+//         not support a reason supported by this platform it will return
+//         RESET_REASON_UNKNOWN.
+reset_reason_t hal_reset_reason_get(void);
```

Comment:
Reading the description, the function seems unreliable. Would it be possible to get the multiple reasons if there is multiple reasons set or at least know that multiple reasons were set. 

It may be a good thing to differentiate an unknown reset reason from an unavailable reset reason.

Reply:
Yes, it's possible on all platforms to have multiple reasons set. Whether it would be better to return each reason ORed together or to return something like `RESET_REASON_MULTIPLE` I'm not sure, I think the latter.

I'll document them and add something like `REASON_OTHER` to indicate platform specific reasons that aren't covered by the enum to distinguish them from actual unknown reasons